src/components/Image/Image.tsx

```diff
@@ -0,0 +1,63 @@
+import React, { useEffect } from 'react';
+
+import { Skeleton } from '../Skeleton';
+
+import styles from './Image.module.scss';
+import classNames from 'classnames';
+
+export interface ImageProps {
+  alt: string;
+  src?: string;
+  className?: string;
+  objectFit?: 'contain' | 'cover' | 'fill' | 'none' | 'scale-down';
+  onLoad?: () => void;
+  onError?: () => void;
+}
+
+interface ImageState {
+  isCached: boolean;
+  src: string;
+}
+
+// named Img here to avoid conflict with the HTMLImageElement in useEffect
+const Img = ({ className, objectFit = 'cover', alt, src, onLoad, onError }: ImageProps) => {
+  const [image, setImage] = React.useState<ImageState | undefined>();
+
+  useEffect(() => {
+    setImage(undefined);
+    if (src && Boolean(new URL(src))) {
+      const image = new Image();
+      image.src = src;
+      // if image is cached, set isCached so we can skip animations
+      if (image.complete) {
+        setImage({ isCached: true, src });
+      } else {
+        // else wait until load is complete to set state
+        image.onload = () => {
+          setImage({ isCached: false, src });
+        };
+      }
+    }
+  }, [src]);
+
+  return (
+    <div
+      className={classNames(styles.Container, className)}
+      data-status={image?.src ? 'loaded' : 'loading'}
```

Comment:
How about to write the state like this?

```
interface ImageState {
  isCached: boolean;
  isLoaded: boolean;
}
.....
const [image, setImage] = React.useState<ImageState>();

....
setImage({ isCashed: false, isLoaded: true});

....
data-status={image?.isLoaded? 'loaded' : 'loading'}
```

Reply:
Yeah I'm not sure which is best to be honest. I like that storing the loaded image in state makes it quite explicit what image you're showing.

I renamed `image` to `loadedImage` for better clarity there!